server/sonar-ce/src/main/java/org/sonar/ce/app/StopFlagContainer.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.ce.app;
+
+public class StopFlagContainer {
```

Comment:
While this certainly does the job, it looks strongly like a patch rather than a solution which takes into account the existing design.

I would prefer an interface `ComputeEngineStatus` being introduced, with a single method `getStatus` returning an enum.
Since this status will be consumed by multiple processes, the interface must document that the implementation of the method must be threadsafe (a very important missing in `StopFlagContainer`) .

This interface would be implemented by `ComputeEngineImpl` which already has an internal status.
The type of the internal property would be changed to the enum of `ComputeEngineStatus` in an `AtomicReference` and `ComputeEngineImpl#getStatus()` would just return it.

`InternalCeQueueImpl` would be injected with the `ComputeEngineStatus` interface. This will elegantly read as "`InternalCeQueueImpl` has a dependency on the Compute Engine status".
Internally, `InternalCeQueueImpl` won't poll for new tasks unless the status is `STARTED`.

Finally, no change would be required in `CeServer` anymore (which will remove some untested code).